### PR TITLE
ARROW-461: [Python] Add Python interfaces to DictionaryArray data, pandas interop

### PR DESCRIPTION
**Pull request by wesm**

### CODE REVIEW

**Comment by wesm** [comment]
@xhochy this is ready to go. I had to move around some code in pandas.cc -- some refactoring of that file into multiple files might be in order

**Comment by wesm** [comment]
Example:

```
In [3]: da = pa.DictionaryArray.from_arrays([1,2,0, 1, 2], ['foo', 'bar', 'baz'])

In [4]: da.to_pandas()
Out[4]: 
[bar, baz, foo, bar, baz]
Categories (3, object): [foo, bar, baz]
```

No DictionaryScalar implemented yet, I can do that in this patch if you'd like

**Review by xhochy — APPROVED** [review]
+1, LGTM

File: python/pyarrow/array.pyx [hunk 1]
```diff
@@ -299,83 +398,29 @@ cdef object box_arrow_array(const shared_ptr[CArray]& sp_array):
...
+        result = pd.Series(obj).values
 
     return result
 
+
 from_pylist = Array.from_list
+from_pandas_series = Array.from_pandas
+
+
+def dictionary(DataType index_type, Array dictionary):
```
Comment (wesm) [inline python/pyarrow/array.pyx]: Yeah, some reorganization with these factory functions may be in order at some point

Comment (wesm) [inline python/pyarrow/array.pyx]: I was able to move it back because `cimport` does not cause a circular import issue

**Pull request closed without merging**